python/ray/tune/function_runner.py

```diff
@@ -51,33 +63,50 @@ def __call__(self, **kwargs):
             StopIteration: A StopIteration exception is raised if the trial has
                 been signaled to stop.
         """
-
         assert self._last_report_time is not None, (
             "StatusReporter._start() must be called before the first "
             "report __call__ is made to ensure correct runtime metrics.")
 
-        # time per iteration is recorded directly in the reporter to ensure
-        # any delays in logging results aren't counted
+        # Time per iteration is recorded directly in the reporter to ensure
+        # any delays in logging results aren't counted.
         report_time = time.time()
         if TIME_THIS_ITER_S not in kwargs:
             kwargs[TIME_THIS_ITER_S] = report_time - self._last_report_time
         self._last_report_time = report_time
 
-        # add results to a thread-safe queue
+        # Add result to a thread-safe queue.
         self._queue.put(kwargs.copy(), block=True)
+        self._wait_for_notification()
 
-        # This blocks until notification from the FunctionRunner that the last
-        # result has been returned to Tune and that the function is safe to
-        # resume training.
-        self._continue_semaphore.acquire()
+    def on_result(self, result):
+        self.__call__(**result)
 
-    def _start(self):
+    def start(self):
         self._last_report_time = time.time()
 
+    def is_pending_restore(self):
```

Comment:
not sure if these functions really belong here but it was the simplest place to put them considering how the code is currently structured